### ISSUE TITLE
CDN (Content Delivery Network) Support

### ISSUE THREAD

**Issue by denydias** [getpelican/pelican#1600]
Hi all,

I'm in the process to deploy many improvements to a work project for which I'm using my own Pelican blog as a lab. One of these improvements is to use a CDN  (AWS CloudFront in that particular case) to distribute static content from the website.

In a typical CDN setup, one set the edge servers to get the static content from an origin server. This origin could be anything, but for the sake of this matter, let's say it's just a typical HTTP server.

The edge server can have a different domain from the origin, lets say:

```
origin: pelican.example.net
cdn: foo.cdn.net
```

Let's also say that one need to CDN deliver only the contents from `$OUTPUTDIR/images` and `$OUTPUTDIR/theme`. No html files delivery for this example.

A way to get tthe assets in `$OUTPUTDIR/theme` delivered over the CDN domain is to edit theme files and hardcode `foo.cdn.net` to each asset (css, js and anything in themes). This works, although it's bad for development as you are going to work with CDN cached assets, so the changes you make will not be promptly applied as you have to wait CDN TTL to expire.

As for the `$OUTPUTDIR/images`, `SITEURL` setting forces the domain to compose the articles image URLs. There is no easy fix for this one.

The goal here is to have the whole Pelican site delivered from `pelican.example.net` (to the world that go straight to this URL and to the CDN itself) and, at the same time, force some assets from `pelican.example.net` to be delivered through the `foo.cdn.net` CDN domain to offload the origin server and improve user experience.

So, now I ask:

1. Is there a way to solve this puzzle within Pelican?
2. Does anybody out there have tried to deliver parts of a Pelican powered site over a CDN provider?
3. How to deal with that multiple URL FQDN?
4. Is Pelican even aware of these requirements for CDN operation?

**Comment by smartass101** [comment]
Pelican was not designed for CDN operation, it assumes it is to be hosted just on one site. For links that are created in the templates, you could use a macro that would inspect the link to be created and change it according to some rules. As for article images, the `pelican.content.Content._update_content` method which modifies the result from the readers would have to be changed.

Another way to do this would be to make a plugin that would hook into the `content_written` signal and would parse the resulting HTML file (e.g. using BeautifulSoup), modifying the links according to some rules. This approach would be simpler to implement at the current moment, but is more of a workaround.

**Comment by denydias** [comment]
Tks for the explanation, @smartass101.

Looking for insights to this, I have found [this issue](https://github.com/DandyDev/pelican-bootstrap3) from `pelican-bootstrap3` theme. This is pretty much the desired behavior, although with some differences, for a nice, development aware CDN integration.

For now, it's pretty clear that Pelican is not designed to do it. That's exactly why I've opened this issue in the hope it gets the proper bits to fulfill this use case, which are becoming pretty common in the past years.

From my pov, I think that at this moment, as an introductory functionality, it would suffice to let Pelican feeds a string set for a hypothetic `CDNURL` in case it exists, for instance, only in `publishconf.py` to replace assets URLs within `make html` and `make publish` steps. Of course, this `CDNURL` could exists in both `pelicanconf.py` and `publishconf.py`, even that in the first it's just `CDNURL = SITEURL`.

This is going to be enough for assets in `themes`, mainly for those Pelican blogs in which `index.html` is not overcrowded by post images and uses just a couple of template images (typically a logo and such). It's going to work just fine for all other js, css and fonts assets that may get loaded from `themes`.

Now that I've put it written, I think that this approach could work as Pelican is designed now. What do you think, @smartass101?

As for the article images, this is going to be a bit more complex as there are plugins that relies on Pelican core functionality. This might be a topic for future discussions.

Unfortunately, I'm not fluent in Python, let alone Pelican. I would love to get into it and code it myself.

**Comment by denydias** [comment]
I've spent the day investigating this issue further. My discoveries go well beyond my initial desires!

As I'm using nginx with ngx_pagespeed, it can deal with all the optimizations I used to apply in Pelican `make publish` time. So, after remove all those optimizers (mainly assets plugin) and configure some bits under nginx, I'm able to rewrite all the URLs I want, which includes css, js and post images!

If someone want to see the results, just head to the URLs bellow and check the assets URLs and headers:

* Home: http://mexapi.macpress.com/
* Article with images: http://mexapi.macpress.com.br/2014/10/livros-livres.html

GTmetrix grades for `/index.html` with this setup: 97% for PageSpeed, 98% for YSlow. A nice double A! Works like a charm for me!

So, this is a non issue for Pelican.
